# kdevplatform.documentation: HelpViewer::loadResource called with unsupported type 2 name= QUrl("qthelp://com.trolltech.qt.487/qdoc/images/qml-webbrowser-demo-small.png")
# kdevelop.plugins.qthelp: cannot determine the content of the new url QUrl("http://www.qt-project.org")
# kdevelop.plugins.qthelp: cannot determine the content of the new url QUrl("http://en.gitorious.org/privacy_policy/")

diff --git kdevplatform/documentation/CMakeLists.txt kdevplatform/documentation/CMakeLists.txt
index 65b94149505b43d2f64041557284635a4c941d0a..732d8d80302f51fc82355d3bd60e5dd5016d9801 100644
--- kdevplatform/documentation/CMakeLists.txt
+++ kdevplatform/documentation/CMakeLists.txt
@@ -8,11 +8,17 @@ if(TARGET Qt5::WebEngineWidgets)
         TYPE REQUIRED)
 else()
     find_package(Qt5WebKitWidgets CONFIG)
-    set_package_properties(Qt5WebKitWidgets PROPERTIES
-        PURPOSE "QtWebKit, for integrated documentation"
-        URL "http://qt-project.org/"
-        TYPE REQUIRED)
-    set(USE_QTWEBKIT 1)
+    if(TARGET Qt5::WebKitWidgets)
+        set_package_properties(Qt5WebKitWidgets PROPERTIES
+            PURPOSE "QtWebKit, for integrated documentation"
+            URL "http://qt-project.org/"
+            TYPE REQUIRED)
+        set(USE_QTWEBKIT 1)
+    else()
+        # store KDEVELOP_USE_QTEXTBROWSER in the cache because it needs to be accessible
+        # to plugins/qthelp/CMakeLists.txt too. 
+        set(KDEVELOP_USE_QTEXTBROWSER 1 CACHE INTERNAL "Documentation will be rendered using QTextBrowser")
+    endif()
 endif()
 
 set(KDevPlatformDocumentation_LIB_SRCS
@@ -32,7 +38,10 @@ kdevplatform_add_library(KDevPlatformDocumentation SOURCES ${KDevPlatformDocumen
 
 target_link_libraries(KDevPlatformDocumentation PUBLIC KDev::Interfaces PRIVATE KDev::Util)
 
-if(USE_QTWEBKIT)
+if(KDEVELOP_USE_QTEXTBROWSER)
+    target_link_libraries(KDevPlatformDocumentation PRIVATE Qt5::Widgets)
+    target_compile_definitions(KDevPlatformDocumentation PRIVATE -DUSE_QTEXTBROWSER)
+elseif(USE_QTWEBKIT)
     target_link_libraries(KDevPlatformDocumentation PRIVATE Qt5::WebKitWidgets)
     target_compile_definitions(KDevPlatformDocumentation PRIVATE -DUSE_QTWEBKIT)
 else()
diff --git kdevplatform/documentation/standarddocumentationview.cpp kdevplatform/documentation/standarddocumentationview.cpp
index 64680e5faaefb0c48f89997be1d9e9198d94e398..0090c93df43bc09882b3ac31e2bab0c40b8ed3a8 100644
--- kdevplatform/documentation/standarddocumentationview.cpp
+++ kdevplatform/documentation/standarddocumentationview.cpp
@@ -32,7 +32,9 @@
 #include <QContextMenuEvent>
 #include <QMenu>
 
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+#include <QTextBrowser>
+#elif defined(USE_QTWEBKIT)
 #include <QFontDatabase>
 #include <QWebView>
 #include <QWebFrame>
@@ -50,7 +52,7 @@
 
 using namespace KDevelop;
 
-#ifndef USE_QTWEBKIT
+#if !defined(USE_QTWEBKIT) && !defined(USE_QTEXTBROWSER)
 class StandardDocumentationPage : public QWebEnginePage
 {
 public:
@@ -80,13 +82,81 @@ private:
 };
 #endif
 
+#ifdef USE_QTEXTBROWSER
+class HelpViewer : public QTextBrowser
+{
+    Q_OBJECT
+public:
+    HelpViewer(StandardDocumentationView* parent)
+        : QTextBrowser(parent)
+        , m_parent(parent)
+        , m_loadFinished(false)
+    {}
+
+    void setSource(const QUrl& url) override
+    {
+        m_loadFinished = false;
+        QTextBrowser::setSource(url);
+    }
+
+    void setUrlWithContent(const QUrl& url, const QByteArray& content)
+    {
+        m_requested = url;
+        m_content = content;
+    }
+
+    // adapted from Qt's assistant
+    QVariant loadResource(int type, const QUrl &name) override
+    {
+        if (type == QTextDocument::HtmlResource) {
+            if (name == m_requested) {
+                qCDebug(DOCUMENTATION) << "loadResource type" << type << "url" << name << "cached=" << m_requested;
+            } else {
+                m_loadFinished = false;
+                emit m_parent->linkClicked(name);
+            }
+        } else if (type != QTextDocument::StyleSheetResource) {
+            qCWarning(DOCUMENTATION) << "HelpViewer::loadResource called with unsupported type" << type << "name=" << name;
+        }
+        // always just return the cached content
+        return m_content;
+    }
+
+    StandardDocumentationView* m_parent;
+    QUrl m_requested;
+    QByteArray m_content;
+    bool m_loadFinished;
+
+Q_SIGNALS:
+    void loadFinished(const QUrl& url);
+
+public Q_SLOTS:
+    void setLoadFinished(bool ok)
+    {
+        m_loadFinished = ok;
+        emit loadFinished(source());
+    }
+};
+
+// void HelpViewer::setLoadFinished(bool ok)
+
+#endif
+
 class KDevelop::StandardDocumentationViewPrivate
 {
 public:
     ZoomController* m_zoomController = nullptr;
     IDocumentation::Ptr m_doc;
 
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    HelpViewer *m_view = nullptr;
+    void init(StandardDocumentationView* parent)
+    {
+        m_view = new HelpViewer(parent);
+        m_view->setContextMenuPolicy(Qt::NoContextMenu);
+        parent->connect(m_view, &HelpViewer::loadFinished, parent, &StandardDocumentationView::linkClicked);
+    }
+#elif defined(USE_QTWEBKIT)
     QWebView *m_view = nullptr;
     void init(StandardDocumentationView* parent)
     {
@@ -130,7 +200,8 @@ StandardDocumentationView::StandardDocumentationView(DocumentationFindWidget* fi
     connect(findWidget, &DocumentationFindWidget::searchDataChanged, this, &StandardDocumentationView::searchIncremental);
     connect(findWidget, &DocumentationFindWidget::searchFinished, this, &StandardDocumentationView::finishSearch);
 
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+#elif defined(USE_QTWEBKIT)
     QFont sansSerifFont = QFontDatabase::systemFont(QFontDatabase::GeneralFont);
     QFont monospaceFont = QFontDatabase::systemFont(QFontDatabase::FixedFont);
 
@@ -176,7 +247,9 @@ KDevelop::StandardDocumentationView::~StandardDocumentationView() = default;
 
 void StandardDocumentationView::search ( const QString& text, DocumentationFindWidget::FindOptions options )
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    typedef QTextDocument WebkitThing;
+#elif defined(USE_QTWEBKIT)
     typedef QWebPage WebkitThing;
 #else
     typedef QWebEnginePage WebkitThing;
@@ -188,12 +261,18 @@ void StandardDocumentationView::search ( const QString& text, DocumentationFindW
     if(options & DocumentationFindWidget::MatchCase)
         ff |= WebkitThing::FindCaseSensitively;
 
+#ifdef USE_QTEXTBROWSER
+    d->m_view->find(text, ff);
+#else
     d->m_view->page()->findText(text, ff);
+#endif
 }
 
 void StandardDocumentationView::searchIncremental(const QString& text, DocumentationFindWidget::FindOptions options)
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    typedef QTextDocument WebkitThing;
+#elif defined(USE_QTWEBKIT)
     typedef QWebPage WebkitThing;
 #else
     typedef QWebEnginePage WebkitThing;
@@ -203,6 +282,9 @@ void StandardDocumentationView::searchIncremental(const QString& text, Documenta
     if (options & DocumentationFindWidget::MatchCase)
         findFlags |= WebkitThing::FindCaseSensitively;
 
+#ifdef USE_QTEXTBROWSER
+    d->m_view->find(text, findFlags);
+#else
     // calling with changed text with added or removed chars at end will result in current
     // selection kept, if also matching new text
     // behaviour on changed case sensitivity though is advancing to next match even if current
@@ -213,12 +295,17 @@ void StandardDocumentationView::searchIncremental(const QString& text, Documenta
     // casesensitivity, that global matches are not updated and the ones with non-matching casing
     // still active. no workaround so far.
     d->m_view->page()->findText(text, findFlags);
+#endif
 }
 
 void StandardDocumentationView::finishSearch()
 {
     // passing emptry string to reset search, as told in API docs
+#ifdef USE_QTEXTBROWSER
+    d->m_view->find(QString());
+#else
     d->m_view->page()->findText(QString());
+#endif
 }
 
 void StandardDocumentationView::initZoom(const QString& configSubGroup)
@@ -253,7 +340,10 @@ void StandardDocumentationView::update()
 
 void KDevelop::StandardDocumentationView::setOverrideCss(const QUrl& url)
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    Q_UNUSED(url);
+    return;
+#elif defined(USE_QTWEBKIT)
     d->m_view->settings()->setUserStyleSheetUrl(url);
 #else
     d->m_view->page()->runJavaScript(QLatin1String(
@@ -269,23 +359,33 @@ void KDevelop::StandardDocumentationView::setOverrideCss(const QUrl& url)
 
 void KDevelop::StandardDocumentationView::load(const QUrl& url)
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    d->m_view->setSource(url);
+#elif defined(USE_QTWEBKIT)
     d->m_view->load(url);
 #else
     d->m_view->page()->load(url);
 #endif
 }
 
+#ifdef USE_QTEXTBROWSER
+void KDevelop::StandardDocumentationView::load(const QUrl& url, const QByteArray& content)
+{
+    d->m_view->setUrlWithContent(url, content);
+    d->m_view->setSource(url);
+}
+#endif
+
 void KDevelop::StandardDocumentationView::setHtml(const QString& html)
 {
-#ifdef USE_QTWEBKIT
+#if defined(USE_QTWEBKIT) || defined(USE_QTEXTBROWSER)
     d->m_view->setHtml(html);
 #else
     d->m_view->page()->setHtml(html);
 #endif
 }
 
-#ifndef USE_QTWEBKIT
+#if !defined(USE_QTWEBKIT) && !defined(USE_QTEXTBROWSER)
 class CustomSchemeHandler : public QWebEngineUrlSchemeHandler
 {
 public:
@@ -306,7 +406,10 @@ private:
 
 void KDevelop::StandardDocumentationView::setNetworkAccessManager(QNetworkAccessManager* manager)
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    Q_UNUSED(manager);
+    return;
+#elif defined(USE_QTWEBKIT)
     d->m_view->page()->setNetworkAccessManager(manager);
 #else
     d->m_view->page()->profile()->installUrlSchemeHandler("qthelp", new CustomSchemeHandler(manager, this));
@@ -315,7 +418,10 @@ void KDevelop::StandardDocumentationView::setNetworkAccessManager(QNetworkAccess
 
 void KDevelop::StandardDocumentationView::setDelegateLinks(bool delegate)
 {
-#ifdef USE_QTWEBKIT
+#ifdef USE_QTEXTBROWSER
+    Q_UNUSED(delegate);
+    return;
+#elif defined(USE_QTWEBKIT)
     d->m_view->page()->setLinkDelegationPolicy(delegate ? QWebPage::DelegateAllLinks : QWebPage::DontDelegateLinks);
 #else
     d->m_page->setLinkDelegating(delegate);
@@ -324,6 +430,9 @@ void KDevelop::StandardDocumentationView::setDelegateLinks(bool delegate)
 
 QMenu* StandardDocumentationView::createStandardContextMenu()
 {
+#ifdef USE_QTEXTBROWSER
+    return d->m_view->createStandardContextMenu();
+#else
     auto menu = new QMenu(this);
 #ifdef USE_QTWEBKIT
     typedef QWebPage WebkitThing;
@@ -336,11 +445,12 @@ QMenu* StandardDocumentationView::createStandardContextMenu()
         menu->addAction(copyAction);
     }
     return menu;
+#endif // !USE_QTEXTBROWSER
 }
 
 bool StandardDocumentationView::eventFilter(QObject* object, QEvent* event)
 {
-#ifndef USE_QTWEBKIT
+#if defined(USE_QTWEBKIT) && !defined(USE_QTEXTBROWSER)
     if (object == d->m_view) {
         // help QWebEngineView properly behave like expected as if Qt::NoContextMenu was set
         if (event->type() == QEvent::ContextMenu) {
@@ -367,7 +477,20 @@ void StandardDocumentationView::contextMenuEvent(QContextMenuEvent* event)
 
 void StandardDocumentationView::updateZoomFactor(double zoomFactor)
 {
+#ifdef USE_QTEXTBROWSER
+    double fontSize = d->m_view->font().pointSizeF();
+    if (fontSize <= 0) {
+        return;
+    }
+    double newSize = fontSize * zoomFactor;
+    if (newSize > fontSize) {
+        d->m_view->zoomIn(int(newSize - fontSize + 0.5));
+    } else if (newSize != fontSize) {
+        d->m_view->zoomOut(int(fontSize - newSize + 0.5));
+    }
+#else
     d->m_view->setZoomFactor(zoomFactor);
+#endif
 }
 
 void StandardDocumentationView::keyPressEvent(QKeyEvent* event)
@@ -385,3 +508,5 @@ void StandardDocumentationView::wheelEvent(QWheelEvent* event)
     }
     QWidget::wheelEvent(event);
 }
+
+#include "standarddocumentationview.moc"
diff --git kdevplatform/documentation/standarddocumentationview.h kdevplatform/documentation/standarddocumentationview.h
index eb5505fb0b14de118e30200de9676d4e4121ee15..ea7a065a2c3d1a20b4fab315f3eaed8216c13cf6 100644
--- kdevplatform/documentation/standarddocumentationview.h
+++ kdevplatform/documentation/standarddocumentationview.h
@@ -59,6 +59,9 @@ public:
     void setOverrideCss(const QUrl &url);
 
     void load(const QUrl &url);
+#ifdef USE_QTEXTBROWSER
+    void load(const QUrl &url, const QByteArray& content);
+#endif
     void setHtml(const QString &html);
     void setNetworkAccessManager(QNetworkAccessManager* manager);
 
diff --git plugins/qthelp/CMakeLists.txt plugins/qthelp/CMakeLists.txt
index 5a8e6e2d1d3677fd590983d2144c10004fffa184..2c5ec0c905a72bfb02325f4eb6c3255aad6350bd 100644
--- plugins/qthelp/CMakeLists.txt
+++ plugins/qthelp/CMakeLists.txt
@@ -28,6 +28,11 @@ target_link_libraries(kdevqthelp
     KF5::KCMUtils KF5::I18n KF5::KIOWidgets KF5::TextEditor KF5::IconThemes Qt5::Help KF5::NewStuff
     KDev::Language KDev::Documentation KDev::Interfaces)
 
+if(KDEVELOP_USE_QTEXTBROWSER)
+    message(STATUS "QtHelp plugin will be built for rendering using QTextBrowser")
+    target_compile_definitions(kdevqthelp PRIVATE -DUSE_QTEXTBROWSER)
+endif()
+
 if(BUILD_TESTING)
     add_subdirectory(tests)
 endif()
diff --git plugins/qthelp/tests/CMakeLists.txt plugins/qthelp/tests/CMakeLists.txt
index 06c751e8a2c254ba175e827fe9036fe635105d4c..2a60e0de2344aafdc040ec207ec0f861991aa7c6 100644
--- plugins/qthelp/tests/CMakeLists.txt
+++ plugins/qthelp/tests/CMakeLists.txt
@@ -24,3 +24,6 @@ ecm_add_test(${test_qthelpplugin_SRCS}
     TEST_NAME test_qthelpplugin
     LINK_LIBRARIES Qt5::Test KF5::NewStuff KF5::KIOWidgets KF5::TextEditor KF5::IconThemes Qt5::Help KDev::Tests KDev::Documentation
 )
+if(KDEVELOP_USE_QTEXTBROWSER)
+    target_compile_definitions(test_qthelpplugin PRIVATE -DUSE_QTEXTBROWSER)
+endif()
diff --git plugins/qthelp/qthelpdocumentation.cpp plugins/qthelp/qthelpdocumentation.cpp
index 19613339d01531b1f7716a9a8e0c2f212bdc856c..a979d0477fbc58fbac2704c92eb47c645f717d90 100644
--- a/plugins/qthelp/qthelpdocumentation.cpp
+++ b/plugins/qthelp/qthelpdocumentation.cpp
@@ -30,6 +30,10 @@
 #include <QTemporaryFile>
 #include <QRegularExpression>
 
+#ifdef USE_QTEXTBROWSER
+#include <QDesktopServices>
+#endif
+
 #include <KLocalizedString>
 
 #include <interfaces/icore.h>
@@ -96,6 +100,18 @@ int lastIndexOf(const QString &str, const QRegularExpression &re, int from, QReg
 }
 #endif
 
+// adapted from Qt's Assistant
+static bool isLocalUrl(const QUrl& url)
+{
+    const QString& scheme = url.scheme();
+    return scheme.isEmpty()
+        || scheme == QLatin1String("file")
+        || scheme == QLatin1String("qrc")
+        || scheme == QLatin1String("data")
+        || scheme == QLatin1String("qthelp")
+        || scheme == QLatin1String("about");
+}
+
 }
 
 QtHelpProviderAbstract* QtHelpDocumentation::s_provider=nullptr;
@@ -224,20 +240,30 @@ QString QtHelpDocumentation::description() const
 
 void QtHelpDocumentation::setUserStyleSheet(StandardDocumentationView* view, const QUrl& url)
 {
+#ifdef USE_QTEXTBROWSER
+    QString css;
+    QTextStream ts(&css);
+#else
     QTemporaryFile* file = new QTemporaryFile(view);
     file->open();
 
     QTextStream ts(file);
+#endif
+
     ts << "html { background: white !important; }\n";
     if (url.scheme() == QLatin1String("qthelp") && url.host().startsWith(QLatin1String("com.trolltech.qt."))) {
        ts << ".content .toc + .title + p { clear:left; }\n"
           << "#qtdocheader .qtref { position: absolute !important; top: 5px !important; right: 0 !important; }\n";
     }
+#ifdef USE_QTEXTBROWSER
+    view->setHtml(css);
+#else
     file->close();
     view->setOverrideCss(QUrl::fromLocalFile(file->fileName()));
 
     delete m_lastStyleSheet.data();
     m_lastStyleSheet = file;
+#endif
 }
 
 QWidget* QtHelpDocumentation::documentationWidget(DocumentationFindWidget* findWidget, QWidget* parent)
@@ -250,12 +276,24 @@ QWidget* QtHelpDocumentation::documentationWidget(DocumentationFindWidget* findW
         view->setDelegateLinks(true);
         view->setNetworkAccessManager(m_provider->networkAccess());
         view->setContextMenuPolicy(Qt::CustomContextMenu);
-        QObject::connect(view, &StandardDocumentationView::linkClicked, this, &QtHelpDocumentation::jumpedTo);
         connect(view, &StandardDocumentationView::customContextMenuRequested, this, &QtHelpDocumentation::viewContextMenuRequested);
 
         setUserStyleSheet(view, m_current.value());
+#ifdef USE_QTEXTBROWSER
+        const auto url = m_current.value();
+        if (isLocalUrl(url)) {
+            view->load(url, m_provider->engine()->fileData(url));
+        } else {
+            // external link
+            qCWarning(QTHELP) << "Opening url" << url << "in the registered web browser";
+            QDesktopServices::openUrl(url);
+        }
+#else
         view->load(m_current.value());
+#endif
         lastView = view;
+        // jumpedTo can only be called safely now.
+        QObject::connect(view, &StandardDocumentationView::linkClicked, this, &QtHelpDocumentation::jumpedTo);
         return view;
     }
 }
@@ -292,7 +330,22 @@ void QtHelpDocumentation::jumpedTo(const QUrl& newUrl)
     Q_ASSERT(lastView);
     m_provider->jumpedTo(newUrl);
     setUserStyleSheet(lastView, newUrl);
+#ifdef USE_QTEXTBROWSER
+    if (isLocalUrl(newUrl)) {
+        QByteArray content = m_provider->engine()->fileData(newUrl);
+        if (!content.isEmpty()) {
+            lastView->load(newUrl, content);
+        } else {
+            qCWarning(QTHELP) << "cannot determine the content of the new url" << newUrl;
+        }
+    } else {
+        // external link, use the user's webbrowser
+        qCWarning(QTHELP) << "Opening new url" << newUrl << "in the registered web browser";
+        QDesktopServices::openUrl(newUrl);
+    }
+#else
     lastView->load(newUrl);
+#endif
 }
 
 IDocumentationProvider* QtHelpDocumentation::provider() const
@@ -309,6 +362,8 @@ QtHelpAlternativeLink::QtHelpAlternativeLink(const QString& name, const QtHelpDo
 void QtHelpAlternativeLink::showUrl()
 {
     IDocumentation::Ptr newDoc(new QtHelpDocumentation(mName, mDoc->info(), mName));
+    // probe, not to be committed.
+    qInfo() << Q_FUNC_INFO << "name" << mName << ":" << mDoc->info();
     ICore::self()->documentationController()->showDocumentation(newDoc);
 }
 
